File: moseoh/_11주차/롤케이크_자르기/java/Solution.java
package _11주차.롤케이크_자르기.java;

import java.util.HashSet;
import java.util.Set;

class Solution {
    public static void main(String[] args) {
        int[] topping1 = new int[]{1, 2, 1, 3, 1, 4, 1, 2};
        int answer1 = 2;
        int result1 = new Solution().solution(topping1);
        PRINT_RESULT(1, result1, answer1);

        int[] topping2 = new int[]{1, 2, 3, 1, 4};
        int answer2 = 0;
        int result2 = new Solution().solution(topping2);
        PRINT_RESULT(2, result2, answer2);
    }

    public static void PRINT_RESULT(int index, int result, int answer) {
        boolean correct = result == answer;
        StringBuilder sb = new StringBuilder();
        sb.append("\n\n테스트 케이스 ").append(index).append(": ");
        sb.append(correct ? "정답" : "오답").append("\n");
        sb.append("\t- 실행 결과: \t").append(result).append("\n");
        sb.append("\t- 기댓값: \t").append(answer).append("\n");
        if (correct) System.out.println(sb);
        else throw new RuntimeException(sb.toString());
    }

    public int solution(int[] topping) {
        int answer = 0;
        Set<Integer> left = new HashSet<>();
        int[] array = new int[topping.length + 1];
        int rightSize = 0;
        for (int top : topping) {
            if (array[top]++ == 0) rightSize++;
        }

        for (int top : topping) {
            left.add(top);
            if (--array[top] == 0) rightSize--;
            if (left.size() == rightSize) answer++;
        }

        return answer;
    }
}

File: moseoh/_11주차/롤케이크_자르기/kotlin/Solution.kt
package _11주차.롤케이크_자르기.kotlin

fun main() {
    fun printResult(index: Int, result: Int, answer: Int) {
        val correct = result == answer
        val sb = StringBuilder()
        sb.append("\n\n테스트 케이스 ").append(index).append(": ")
        sb.append(if (correct) "정답" else "오답").append("\n")
        sb.append("\t- 실행 결과: \t").append(result).append("\n")
        sb.append("\t- 기댓값: \t").append(answer).append("\n")
        if (correct) println(sb) else throw RuntimeException(sb.toString())
    }

    val topping1 = intArrayOf(1, 2, 1, 3, 1, 4, 1, 2)
    val answer1 = 2
    val result1 = Solution().solution(topping1)
    printResult(1, result1, answer1)

    val topping2 = intArrayOf(1, 2, 3, 1, 4)
    val answer2 = 0
    val result2 = Solution().solution(topping2)
    printResult(2, result2, answer2)
}

class Solution {
    fun solution(topping: IntArray): Int {
        var answer = 0
        val leftSet = mutableSetOf<Int>()
        val array = IntArray(topping.size + 1)
        var rightSize = 0
        topping.forEach { if (array[it]++ == 0) rightSize++ }

        topping.forEach {
            leftSet.add(it)
            if (--array[it] == 0) rightSize--
            if (leftSet.size == rightSize) {
                answer++
            }
        }

        return answer
    }
}